docs/proposals/web-hook.md

```diff
@@ -0,0 +1,105 @@
+
+There is a concrete example on the github [issue 845](https://github.com/kubeedge/kubeedge/issues/845), the issue there cannot be addressed by [CRD validation](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#validation), so we need to explore approach to achieve that purpose.
+
+
+## Goals
+* A framework to enable admission control in kubeedge.
+* Address the edge cases where kubernetes CRD validation cannot accomplish.
+* Create basic integration testcases.
+
+
+# Proposal
+Propose using Kubernetes [Dynamic Admission Control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers) to determine whether to accept or deny the request based on how is the policy is configured, this give us chance to validate the request before persisting the object.
+
+
+# Design Details
+
+## Admission controller on control-plane
+
+Create a new admission control when the module `controller` is started.
+```go
+func (ctl *Controller) Start(c *bcontext.Context) {
+	config.Context = c
+	upstream, err := controller.NewUpstreamController()
+	...
+        admission, err := controller.NewAdmissionController()
+	...
+        admission.Start()
+        ...
+}
+```
+
+
+## Start a http Server
+
+The http server will listen on the specific port, `8899` for example, the http server will start to listen when the admission controller is started.
+```go
+server := &http.Server{
+	Handler:   ac,
+	Addr:      constants.Port,
+	TLSConfig: tlsConfig,
+}
+```
+
+`ServeHTTP` will be used to implement the admission webhook for resource validation.
+```go
+func (ac *AdmissionController) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	...
+	if ar.Request.Operation == v1beta1.create {
+		//enforce custom policy
+		reviewResponse.Allowed = true
+	}
+	...
+	return &reviewResponse
+}
+```
+
+## Register admission webhook for each configuration
+Rules could be centralized here, as an example, register one admission webhook as below,
+
+```go
+Webhooks: []admissionregistrationv1beta1.ValidatingWebhook{
+	{
+		Name: "deny-unwanted-pod-container-name-and-label.k8s.io",
+		Rules: []admissionregistrationv1beta1.RuleWithOperations{{
+			Operations: []admissionregistrationv1beta1.OperationType{admissionregistrationv1beta1.Create},
+			Rule: admissionregistrationv1beta1.Rule{
+				APIGroups:   []string{""},
+				APIVersions: []string{"v1"},
+				Resources:   resources,
+			},
+		}},
+		ClientConfig: admissionregistrationv1beta1.WebhookClientConfig{
+			Service: &admissionregistrationv1beta1.ServiceReference{
+				Namespace: namespace,
+				Name:      serviceName,
+				Port:      pointer.Int32Ptr(servicePort),
+			},
+			CABundle: context.signingCert,
+		},
+	},
+	...
+	}
+```
+
+
+## Certification management
+Need to generate another set of key / certification, or reuse the existing key / certification for API validation.
```

Comment:
Can we refer from here : https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#authenticate-apiservers